pkg/v1/v1util/matcher.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1util
+
+import (
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ManifestMatcher function that is given a v1.Descriptor, and returns whether or
+// not it matches a given rule. Can match on anything it wants in the Descriptor.
+type ManifestMatcher func(desc v1.Descriptor) bool
+
+// NameMatcher returns a ManifestMatcher that matches on the name tag in
+// github.com/opencontainers/image-spec/specs-go/v1.AnnotationRefName
+func NameMatcher(name string) ManifestMatcher {
```

Comment:
Started typing out "what if we wanted to match multiple names?" but it might be better to have an API for composing matchers... e.g.:

```go
And(Or(NameMatcher("foo"), NameMatcher("bar")), Or(PlatformMatcher(arm), PlatformMatcher(arm64)))
```

Which would match anything images with name "foo" or "bar" that also have platform "arm" or "arm64".

```go
func Or(matchers ...ManifestMatcher) ManifestMatcher {
  return func(desc v1.Descriptor) bool {
    for _, m := range matchers {
      if m(desc) {
        return true
      }
    }
    return false
  }
}

func And(matchers ...ManifestMatcher) ManifestMatcher {
  return func(desc v1.Descriptor) bool {
    for _, m := range matchers {
      if !m(desc) {
        return false
      }
    }
    return true
  }
}
```

Now we've got a whole language, and we maybe want a new package for this craziness, e.g.:

```go
match.And(
  match.Or(match.Name("foo"), match.Name("bar")),
  match.Or(match.Platform(arm), match.Platform(arm64))
)
```

Reply:
That is pretty complex. I suspect for the vast majority of use cases, people will be looking just for one name. This is the closest thing to `remote.Get()` I can think of. How about if we start with the simple matchers, and then create composites later?

Reply:
Agreed, we shouldn't start here, but if we want to end up here, this is a good argument for pulling out matchers into a separate matching-only package so to make names a little less verbose and avoid polluting other package namespaces.

Reply:
If that is the approach, I can pull it out into a `matcher` package. I am a little concerned about getting into a circular imports loop (matcher importing v1 importing matcher), but I think we will be ok.

Let me know.

Reply:
> matcher importing v1 importing matcher

This is why `v1` is (almost) strictly structs and interfaces. There's some stuff for JSON [de]serialization in there as well, but it should be as small as possible and have as few dependencies as possible. Any real behavior should go in a different package.